tests/database/odbc-w-enclave/cksp.c

```diff
@@ -0,0 +1,155 @@
+    CEKEYSTORECONTEXT* ctx,
+    errFunc* onError,
+    void* data,
+    unsigned int len)
+{
+    assert(len == sizeof(data));
+    return CKSP_OK;
+}
+
+/***************************************************************************
+  This function is invoked by ODBC driver whenever an ECEK needs to be
+  decrypted, as the result of a query involving encrypted columns.
+
+  Parameters:
+    ctx:      current context of the CSKP
+    onError:  error handler
+    keyPath:  the AKV path to the CMK in wchars
+    alg:      algorithm used to encrypt the CEK
+    ecek:     the encrypted ECK
+    ecekLen:  the length of the ECEK
+    cekOut:   the double pointer to the buffer we allocate and save CEK
+    cekLen:   the length of the CEK
+  ***************************************************************************/
+int __stdcall KeystoreDecrypt(
+    CEKEYSTORECONTEXT* ctx,
+    errFunc* onError,
+    const wchar_t* keyPath,
+    const wchar_t* alg,
+    unsigned char* ecek,
+    unsigned short ecekLen,
+    unsigned char** cekOut,
+    unsigned short* cekLen)
+{
+    if ((*cekOut = malloc(ecekLen)) == NULL)
+    {
+        fprintf(stderr, "KSP Encrypt: out of memory\n");
+        return CKSP_ERROR;
+    }
+
+    *cekLen = ecekLen;
+    // simple decryption.
+    for (size_t i = 0; i < ecekLen; i++)
+        (*cekOut)[i] = ecek[i] - 1;
+
+    fprintf(stderr, "KSP Decrypt(): Successful\n");
+    return CKSP_OK;
+}
+
+/***************************************************************************
+  A client application will call this function to encrypt a given CEK.
+  The output ECEK is then stored in the database.
+
+  Parameters:
+    ctx:      current context of the CSKP
+    onError:  error handler
+    keyPath:  the real key path to the CMK in wchars
+    alg:      algorithm used to encrypt the CEK
+    cek:      the unencrypted ECK
+    cekLen:   the length of the CEK
+    ecekOut:   the double pointer to the buffer we allocate and save ECEK
+    ecekLen:   the length of the ECEK
+  ***************************************************************************/
+int KeystoreEncrypt(
+    CEKEYSTORECONTEXT* ctx,
+    errFunc* onError,
+    const wchar_t* keyPath,
+    const wchar_t* alg,
+    unsigned char* cek,
+    unsigned short cekLen,
+    unsigned char** ecekOut,
+    unsigned short* ecekLen)
+{
+    if ((*ecekOut = malloc(cekLen)) == NULL)
+    {
+        fprintf(stderr, "KSP Encrypt: out of memory\n");
+        return CKSP_ERROR;
+    }
+
+    *ecekLen = cekLen;
+    // simple encryption.
+    for (size_t i = 0; i < cekLen; i++)
+        (*ecekOut)[i] = cek[i] + 1;
+
+    return CKSP_OK;
+}
+
+int VerifyCMKMetadata(
+    CEKEYSTORECONTEXT* ctx,
+    errFunc* onError,
+    const wchar_t* keyPath,
+    unsigned char* signature,
+    unsigned short sigLen)
+{
+    return 1;
+}
+
+void KeyStoreFree()
+{
+}
+
+CEKEYSTOREPROVIDER2 MyCustomKSPName_desc = {
+    // pretend to be AKV, so SQL server will forward the ECEK decryption
```

Comment:
Is this a hack? Will this break in the future?

Reply:
@jxyang can you comment on this?

Reply:
This is a hack suggested by SQL team because AEv2 only recognizes built-in key vaults, not the custom ones. In fact, if we register with AEv2 a custom key vault, it doesn't even forward the key decryption request to the driver. Yes this could break in the future, and we should document  in more details.